http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeoAuthProvider.scala

```diff
@@ -103,6 +110,15 @@ object SamResourceAction {
     def actionAsString(a: AppAction): String = a.asString
   }
 
+  implicit def SharedAppSamResourceAction = new SharedAppSamResource
```

Comment:
I have to admit that I am not sure what this does... It does not seem to be used anywhere else (but I keep getting burned by these implicits)

Reply:
These implicits maybe go overboard a bit. I think basically it's trying to tie together Sam resource types + actions at compile time. So it's impossible to call `hasResourcePermission(resource, action)` for an action that doesn't match the resource type (e.g. it doesn't compile).

… Which is kinda cool, but at the same time I struggle to remember how this code works (and I wrote it!), so it feels like the complexity isn't really worth it.

Reply:
I ended up modifying the SamResource object to have ResourceType and SamPolicies as functions instead of values. This makes things more flexible, and avoid having to pass around the `accessScope` parameter